File: src/main/java/destiny/FengShui/SanYuan/Chart.java
/**
 * @author smallufo
 * @date 2002/9/25
 * @time 上午 02:27:05
 * 
 * 2009/12/15 把與視覺相關的 ColorCanvas 移出到 {@link destiny.FengShui.SanYuan.ChartColorCanvasWrapper} , 讓此 Chart 純粹只有資料結構
 */
package destiny.FengShui.SanYuan;

import destiny.iching.Symbol;
import destiny.iching.SymbolAcquired;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;

/**
 * 三元盤的 presentation model 
 */
public class Chart implements Serializable {
  /**
   * 元運 , 其值只能為 1~9
   */
  private int period = 7; //內訂是七運
  
  /**
   * 座山
   */
  private Mountain mountain = Mountain.子; //內訂是子山
  
  /**
   * 從哪個卦看去 ?
   * 意即，九宮格的底邊，是哪個卦（後天）
   * 內定以坎卦作為底邊，即，底邊為北方
   */
  private Symbol view = Symbol.坎; 
  
  @NotNull
  private ChartBlock[] blocks = new ChartBlock[10]; // 0 不用
  //private ChartBlock[][] chartCoordinate = new ChartBlock[3][3];

  @NotNull
  private EarthlyCompass 地盤 = new EarthlyCompass();

  @NotNull
  private AcquiredSymbolCompass 後天八卦盤 = new AcquiredSymbolCompass();
  
  /**
   * Constructor , 設定年運以及座山 , 內訂觀點為坎
   */
  public Chart(int period , Mountain mountain)
  {
    this.period = period;
    this.mountain = mountain;
    calculate();
  }
  
  /**
   * constructor , 設定年運、座山，以及底邊方向
   */
  public Chart(int period, Mountain mountain, Symbol view) {
    this.period = period;
    this.mountain = mountain;
    this.view = view;
    calculate();
  }
  
  private void calculate() {
    /** 
     * 先將chartBlockList initialize 
     * 設定其宮位 , 該元運置於 blocks[1] , 順推
     */
    int 元運 = getPeriod();
    for (int i = 1; i <= 9; i++) {
      blocks[i] = new ChartBlock(normalize(元運++));
    }

    //                                    |-----------------|
    //設定方向（以卦來表示），尚未實作轉盤   // | 9巽 | 5離 | 7坤 |
    blocks[9].setSymbol(Symbol.巽);    // |[0,0]|[0,1]|[0,2]|
    blocks[8].setSymbol(Symbol.震);    // |-----|-----|-----|
    blocks[4].setSymbol(Symbol.艮);    // | 8震 |  1  | 3兌 |
    blocks[5].setSymbol(Symbol.離);    // |[1,0]|[1,1]|[1,2]|
    blocks[6].setSymbol(Symbol.坎);    // |-----|-----|-----|
    blocks[7].setSymbol(Symbol.坤);    // | 4艮 | 6坎 | 2乾 |
    blocks[3].setSymbol(Symbol.兌);    // |[2,0]|[2,1]|[2,2]|
    blocks[2].setSymbol(Symbol.乾);    // |-----------------|

    //決定座山/向 是位於哪一卦中
    //詢問此山/向 的中心點度數為：
    double midMountain;
    if (Math.abs
       (地盤.getEndDegree  (getMountain()) - 
        地盤.getStartDegree(getMountain()) ) 
       > 180 )
      midMountain = 0;
    else
      midMountain =  
      (地盤.getEndDegree(getMountain()) +
       地盤.getStartDegree(getMountain()) ) / 2;
    
    double midDirection;
    if (midMountain == 180.0)
      midDirection = 0;
    else
      midDirection =(
      地盤.getEndDegree(getMountain().getOppositeMountain()) + 
      地盤.getStartDegree(getMountain().getOppositeMountain()) ) / 2 ;

    Symbol 飛佈山盤卦 = (Symbol) 後天八卦盤.getSymbol(midMountain );
    Symbol 飛佈向盤卦 = (Symbol) 後天八卦盤.getSymbol(midDirection);
    
    //搜尋 blocks[1~9] , 分別找尋 飛佈山盤卦 以及 飛佈向盤卦 , 取得其 period 值
    int mountainStart  = 0;
    int directionStart = 0;
    for (int i=1 ; i <=9 ; i++) {
      if (blocks[i].getSymbol() == 飛佈山盤卦 )
        mountainStart = blocks[i].getPeriod();
      if (blocks[i].getSymbol() == 飛佈向盤卦 )
        directionStart = blocks[i].getPeriod();
    }

    // TODO : 以後改用 SymbolAcquired.getSymbol : Optional<Symbol>  的演算法
    Symbol 原始山盤卦 = SymbolAcquired.getSymbolNullable(mountainStart);
    Symbol 原始向盤卦 = SymbolAcquired.getSymbolNullable(directionStart) ;
    
    //填入山盤
    fillMountain (mountainStart  , isConverse(原始山盤卦 , 飛佈山盤卦 , getMountain()) );
    //填入向盤
    fillDirection(directionStart , isConverse(原始向盤卦 , 飛佈向盤卦 , getMountain().getOppositeMountain())  );
  } //calculate()
  
  
  /**
   * 查詢某卦方位裡面的資料結構 (ChartBlock) , 如果查的是 null , 則傳回中宮
   */
  @Nullable
  public ChartBlock getChartBlock(Symbol s)
  {
    for(ChartBlock chartBlock : blocks)
    {
      if (chartBlock != null && chartBlock.getSymbol() == s)
        return chartBlock;
    }
    throw new RuntimeException("Cannot find ChartBlock " + s);
  }
  
  private boolean isConverse(@Nullable Symbol 原始卦 , Symbol 飛佈卦 , @NotNull Mountain m)
  {
    boolean isConverse = false;
    if (原始卦 == null)
    {
      if (地盤.getYinYang(m))
        isConverse = false;
      else
        isConverse = true;
    }
    else
    {
      int index = (int) ((地盤.getStartDegree(m)- 後天八卦盤.getStartDegree(飛佈卦) ) / 15);
      /*
       * index = 0 => 地元
       * index = 1 => 天元
       * index = 2 => 人元
       */
      double degree = 後天八卦盤.getStartDegree(原始卦) + index*15 +1; //最後的 +1 是確保結果能坐落於該山中
      if (地盤.getYinYang((Mountain) (地盤.getMountain(degree))))
        isConverse = false;
      else
        isConverse = true;
    }
    return isConverse;
  }
  
  /**
   * 將 int 值侷限在 1 到 9 之間
   */
  private static int normalize(int value)
  {
    if (value > 9 )
      return normalize( value-9);
    else if (value < 1)
      return normalize (value+9);
    else     
      return value;
  }
  
  
  /**
   * 填滿山盤，傳入座山 , 以及是否逆飛
   */
  private void fillMountain(int start , boolean converse)
  {
    if (!converse)
    {
      //順飛
      for (int i=1 ; i <=9 ; i++)
      {
        blocks[i].setMountain(normalize(start + i + 8));
      }
    }
    else
    {
      //逆飛
      for (int i=1 ; i <=9 ; i++)
      {
        blocks[i].setMountain(normalize(start - i + 10));
      }
    }
  }//fillMountain
  
  
  /**
   * 填滿向盤，傳入向山 , 以及是否逆飛
   */
  private void fillDirection(int start , boolean converse)
  {
    if (!converse)
    {
      //順飛
      for (int i=1 ; i <=9 ; i++)
      {
        blocks[i].setDirection(normalize(start + i + 8));
      }
    }
    else
    {
      //逆飛
      for (int i=1 ; i <=9 ; i++)
      {
        blocks[i].setDirection(normalize(start - i + 10));
      }
    }
  }//fillMountain
  
  /**
   * 傳回元運
   */
  public int getPeriod() 
  {
    return period; 
  }
  
  /**
   * 傳回座山
   */
  public Mountain getMountain() 
  { 
    return mountain; 
  }
  
  /**
   * 傳回底邊
   */
  public Symbol getView() 
  { 
    return view; 
  }
}

File: src/main/java/destiny/FengShui/SanYuan/Chart.kt
/**
 * Created by smallufo on 2018-02-25.
 */
package destiny.FengShui.SanYuan

import destiny.iching.Symbol
import destiny.iching.Symbol.*
import destiny.iching.SymbolAcquired
import java.io.Serializable

/**
 * 三元盤的 presentation model
 * 挨星下卦
 */
class Chart(

  // 元運 , 其值只能為 1~9
  val period: Int,

  // 座山
  val mountain: Mountain,

  // 從哪個卦看去 ? 意即，九宮格的底邊，是哪個卦（後天）
  val view: Symbol) : Serializable {

  private val blocks = arrayOfNulls<ChartBlock>(10) // 0 不用


  init {

    //決定座山/向 是位於哪一卦中
    //詢問此山/向 的中心點度數為：
    val midMountain: Double = if (Math.abs(地盤.getEndDegree(mountain) - 地盤.getStartDegree(mountain)) > 180)
      0.0
    else
      (地盤.getEndDegree(mountain) + 地盤.getStartDegree(mountain)) / 2

    val midDirection = if (midMountain == 180.0)
      0.0
    else
      (地盤.getEndDegree(mountain.opposite) + 地盤.getStartDegree(mountain.opposite)) / 2

    val 飛佈山盤卦 = 後天八卦盤.getSymbol(midMountain)
    val 飛佈向盤卦 = 後天八卦盤.getSymbol(midDirection)

    //搜尋 blocks[1~9] , 分別找尋 飛佈山盤卦 以及 飛佈向盤卦 , 取得其 period 值
    val mntStart: Int = (1..9).first { getBlockSymbol(it) === 飛佈山盤卦 }.let { getBlockPeriod(it) }
    val dirStart: Int = (1..9).first { getBlockSymbol(it) === 飛佈向盤卦 }.let { getBlockPeriod(it) }

    // 原始 == 中宮
    val 原始山盤卦: Symbol? = SymbolAcquired.getSymbolNullable(mntStart)
    val 原始向盤卦: Symbol? = SymbolAcquired.getSymbolNullable(dirStart)

//    println("${飛佈山盤卦}山 ($midMountain) ${飛佈向盤卦}向")
//    println("\t$飛佈山盤卦 = $mntStart , from 後天八卦 $原始山盤卦")
//    println("\t$飛佈向盤卦 = $dirStart , from 後天八卦 $原始向盤卦")


    val mntConverse = isConverse(原始山盤卦, 飛佈山盤卦, mountain)
    val dirConverse = isConverse(原始向盤卦, 飛佈向盤卦, mountain.opposite)

    for (i in 1..9) {
      val period = normalize(period + i - 1)
      val symbol = getBlockSymbol(i)

      val mnt = getNumber(mntStart, mntConverse, i)
//      println("[$i] , mntStart = $mntStart , mntConverse = $mntConverse")
      val dir = getNumber(dirStart, dirConverse, i)

      blocks[i] = ChartBlock(symbol, mnt, dir, period)
    }
  } //init


  /**
   * 方向（以卦來表示）
   * |-----------------|
   * | 9巽 | 5離 | 7坤 |
   * |[0,0]|[0,1]|[0,2]|
   * |-----|-----|-----|
   * | 8震 |  1  | 3兌 |
   * |[1,0]|[1,1]|[1,2]|
   * |-----|-----|-----|
   * | 4艮 | 6坎 | 2乾 |
   * |[2,0]|[2,1]|[2,2]|
   * |-----------------|
   */
  private fun getBlockSymbol(blockIndex: Int): Symbol? {
    return when (blockIndex) {
      9 -> 巽
      8 -> 震
      4 -> 艮

      5 -> 離
      6 -> 坎

      7 -> 坤
      3 -> 兌
      2 -> 乾
      else -> null
    }
  }

  private fun getBlockPeriod(blockIndex: Int): Int {
    val 元運 = period
    return normalize(元運 + blockIndex - 1)
  }

  /**
   * 查詢某卦方位裡面的資料結構 (ChartBlock)
   */
  fun getChartBlock(s: Symbol): ChartBlock {
    return blocks.first { cb: ChartBlock? -> cb?.symbol === s }!!
  }

  /**
   * 則傳回中宮
   */
  fun getCenterBlock() : ChartBlock {
    //blocks.forEach { println(it) }
    return blocks.filter { it != null }.first { cb -> cb?.symbol === null }!!
  }

  private fun isConverse(原始卦: Symbol?, 飛佈卦: Symbol, m: Mountain): Boolean {
    val isConverse: Boolean
    isConverse = if (原始卦 == null) {
      /**
       * 當五黃如中時，因為五黃本身沒有山向，五黃是中宮戊己土，那麼，
       * 五入中的順飛還是逆飛則是由五所在的宮位的山向的陰陽，性質決定。
       * 如一運子山午向，運星五到離，五入中為向星後，按五所在的「午」的性質決定，故逆飛。
       */
      !地盤.getYinYang(m)
    } else {
      val index = ((地盤.getStartDegree(m) - 後天八卦盤.getStartDegree(飛佈卦)) / 15).toInt()
      /**
       * index = 0 => 地元
       * index = 1 => 天元
       * index = 2 => 人元
       */
      val degree = 後天八卦盤.getStartDegree(原始卦) + (index * 15).toDouble() + 1.0 //最後的 +1 是確保結果能坐落於該山中
      !地盤.getYinYang(地盤.getMnt(degree))
    }
    return isConverse
  }

  /**
   * 將 int 值侷限在 1 到 9 之間
   */
  private fun normalize(value: Int): Int {
    return when {
      value > 9 -> normalize(value - 9)
      value < 1 -> normalize(value + 9)
      else -> value
    }
  }


  /**
   * 傳入中宮的 山 或 向  , 以及是否逆飛
   *
   * @param blockIndex 1 to 9
   */
  private fun getNumber(start: Int, converse: Boolean, blockIndex: Int): Int {
    return if (!converse) {
      // 順飛
      normalize(start + blockIndex + 8)
    } else {
      // 逆飛
      normalize(start - blockIndex + 10)
    }
  }


  companion object {
    private val 地盤 = EarthlyCompass()
    private val 後天八卦盤 = AcquiredSymbolCompass()
  }

}
